### PR TITLE
Refactor fixture loader

### LINKED ISSUE #878: Re-factor test fixture loader
Inserting by index looks extremely brittle. Could we do something like:

```python
from dataclass import fields

@dataclass
class Foobar:
    nonce: Optional[Uint]
    gasLimit: Optional[Uint]
    # ...

    def to_list(self) -> List[?]:
        parameters = []
        for name in fields(self):
            value = getattr(self, name)
            if value is None:
                continue
            parameters.append(value)
```

Or is that overkill?

_Originally posted by @SamWilsn in https://github.com/ethereum/execution-specs/pull/867#discussion_r1475180953_

### PR DESCRIPTION
**Pull request by gurukamath**
(closes #878 )

### What was wrong?

Fixture loader was starting to get bloated and needed to be re-factored.

Related to Issue #878 

### How was it fixed?
Create separate data classes for loading fork code (ForkLoad), json fixture loader (Load) and json transaction loader (TransactionLoad)

#### Cute Animal Picture

Library - 1 of 1

### CODE REVIEW

File: src/ethereum/cancun/fork_types.py [hunk 1]
```diff
@@ -116,7 +116,7 @@ class BlobTransaction:
     max_priority_fee_per_gas: Uint
     max_fee_per_gas: Uint
     gas: Uint
-    to: Union[Bytes0, Address]
+    to: Address
```
Comment (SamWilsn) [inline src/ethereum/cancun/fork_types.py:119]: Should we upstream a test for this?

File: src/ethereum/rlp.py [hunk 2]
```diff
@@ -237,7 +237,7 @@ def _decode_to(cls: Type[T], raw_rlp: RLP) -> T:
             for t, raw_item in zip(cls.__args__, raw_rlp):  # type: ignore
                 args.append(_decode_to(t, raw_item))
             return tuple(args)  # type: ignore
-    elif cls == Union[Bytes0, Bytes20]:
+    elif cls == Union[Bytes0, Bytes20] or cls == Bytes20:
```
Comment (SamWilsn) [inline src/ethereum/rlp.py:240]: This new line doesn't match the subsequent comment. How about something like:

```suggestion
    elif issubclass(cls, FixedBytes):
        ensure(isinstance(raw_rlp, Bytes), RLPDecodingError)
        ensure(len(raw_rlp) == cls.LENGTH, RLPDecodingError)
        return cls(raw_rlp)  # type: ignore
    elif cls == Union[Bytes0, Bytes20]:
```

File: src/ethereum_spec_tools/evm_tools/loaders/fixture_loader.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    hex_to_u256,
+    hex_to_uint,
+)
+
+from .fork_loader import ForkLoad
+from .transaction_loader import TransactionLoad
+
+
+class BaseLoad(ABC):
+    """Base class for loading json fixtures"""
```
Comment (SamWilsn) [inline src/ethereum_spec_tools/evm_tools/loaders/fixture_loader.py:28]:
```suggestion
    "Base class for loading JSON fixtures"
```

Can use one double quote for single line docstrings.

File: src/ethereum_spec_tools/evm_tools/loaders/fixture_loader.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    @abstractmethod
+    def json_to_header(self, json_data: Any) -> Any:
+        """Converts json header data to a header object"""
+        pass
```
Comment (SamWilsn) [inline src/ethereum_spec_tools/evm_tools/loaders/fixture_loader.py:33]:
```suggestion
        "Converts json header data to a header object"
        raise NotImplementedError()
```

Saw this style of abstract method in another Python project. What do you think?

File: src/ethereum_spec_tools/evm_tools/t8n/env.py [hunk 5]
```diff
@@ -66,14 +68,47 @@ def __init__(self, t8n: Any, stdin: Optional[Dict] = None):
...
+                excess_blob_gas = (
+                    self.parent_excess_blob_gas + self.parent_blob_gas_used
+                )
+
+                if excess_blob_gas < t8n.vm_gas.TARGET_BLOB_GAS_PER_BLOCK:
+                    self.excess_blob_gas = U64(0)
+                else:
+                    self.excess_blob_gas = (
+                        excess_blob_gas - t8n.vm_gas.TARGET_BLOB_GAS_PER_BLOCK
+                    )
```
Comment (SamWilsn) [inline src/ethereum_spec_tools/evm_tools/t8n/env.py:111]:
```suggestion
                )
                continue

            self.parent_excess_blob_gas = parse_hex_or_int(
                data["parentExcessBlobGas"], U64
            )
            self.parent_blob_gas_used = parse_hex_or_int(
                data["parentBlobGasUsed"], U64
            )

            excess_blob_gas = (
                self.parent_excess_blob_gas + self.parent_blob_gas_used
            )

            if excess_blob_gas < t8n.vm_gas.TARGET_BLOB_GAS_PER_BLOCK:
                self.excess_blob_gas = U64(0)
            else:
                self.excess_blob_gas = (
                    excess_blob_gas - t8n.vm_gas.TARGET_BLOB_GAS_PER_BLOCK
                )
```

**Review by SamWilsn — COMMENTED** [review]
Only a partial review so far, sorry!

File: src/ethereum_spec_tools/evm_tools/loaders/transaction_loader.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+        """Get the gas limit for the transaction."""
+        return hex_to_u256(self.raw.get("gasLimit"))
+
+    def json_to_to(self) -> Bytes:
+        """Get to address for the transaction."""
+        return (
+            Bytes0(b"")
+            if self.raw.get("to") == ""
+            else self.fork.hex_to_address(self.raw.get("to"))
+        )
```
Comment (SamWilsn) [inline src/ethereum_spec_tools/evm_tools/loaders/transaction_loader.py:59]:
```suggestion
        if self.raw.get("to") == "":
            return Bytes0(b"")
        return self.fork.hex_to_address(self.raw.get("to"))
```

File: tests/spurious_dragon/test_state_transition.py [hunk 7]
```diff
@@ -131,31 +131,31 @@ def test_transaction_with_insufficient_balance_for_value() -> None:
...
         nonce=Uint(0),
         balance=U256(0x056BC75E2D63100000),
         code=Bytes(),
     )
 
-    FIXTURES_LOADER.set_account(state, address, account)
+    FIXTURES_LOADER.fork.set_account(state, address, account)
 
-    tx = FIXTURES_LOADER.LegacyTransaction(
+    tx = FIXTURES_LOADER.fork.Transaction(
```
Comment (SamWilsn) [inline tests/spurious_dragon/test_state_transition.py:154]: Just confirming this change is intentional.

**Review by SamWilsn — COMMENTED** [review]
And that's the rest of it.

File: src/ethereum/cancun/fork_types.py (hunk 1 above)
Comment (gurukamath) [reply src/ethereum/cancun/fork_types.py:119]: There is already a test for this in the latest `ethereum/tests`

File: src/ethereum/rlp.py (hunk 2 above)
Comment (gurukamath) [reply src/ethereum/rlp.py]: Updated

File: src/ethereum_spec_tools/evm_tools/loaders/fixture_loader.py (hunk 4 above)
Comment (gurukamath) [reply src/ethereum_spec_tools/evm_tools/loaders/fixture_loader.py]: Makes sense. Updated

File: src/ethereum_spec_tools/evm_tools/t8n/env.py (hunk 5 above)
Comment (gurukamath) [reply src/ethereum_spec_tools/evm_tools/t8n/env.py]: Updated

**Comment by gurukamath** [comment]
Completed the updates to the PR based on the review and rebased it on latest `forks/cancun`.

File: src/ethereum_spec_tools/evm_tools/loaders/transaction_loader.py (hunk 6 above)
Comment (gurukamath) [reply src/ethereum_spec_tools/evm_tools/loaders/transaction_loader.py]: Updated

**Pull request merged**